### PR TITLE
feat: multi task import

### PR DESCRIPTION
**Pull request by xjlgod**

### CODE REVIEW

File: server/README.md [hunk 1]
```diff
@@ -29,15 +29,19 @@ $ ./nebula-studio-server -studio-config="./config/example-config.yaml"
 | ConnectDB          | /api-nebula/db/connect | POST     |
 | DisconnectDB       | /api-nebula/db/disconnect | POST     |
 | ImportData         | /api-nebula/task/import   | POST     |
-| HandleImportAction | /api-nebula/import/action | POST     |
+| HandleImportAction | /api-nebula/task/import/action | POST     |
+| QueryImportStats | /api-nebula/task/import/stats | POST |
+| DownloadConfig | /api-nebula/task/import/config | POST |
+| DownloadLog | /api-nebula/task/import/log | GET |
```
Comment (veezhang) [inline server/README.md]: Why not use restful style api?

File: server/README.md [hunk 2]
```diff
@@ -119,35 +123,153 @@ Response:
...
+  concurrency: 10
+  channelBufferSize: 128
+  space: test
+  connection:
+    user: user
+    password: password
+    address: host
+  postStart: null
+  preStop: null
+logPath: E:\NebulaProject\新建文件夹 (4)\nebula-studio\server\data\tasks\1/import.log
```
Comment (veezhang) [inline server/README.md]: Do not use Chinese.

File: server/README.md [hunk 3]
```diff
@@ -119,35 +123,153 @@ Response:
...
+          type: int
+          index: 2
+
+```
+
+#### DownloadLog API
+
+The request :
+
+http://localhost:9000/api-nebula/task/import/log/?pathName=E:\NebulaProject\nebula-studio\server\data\tasks\1\err\err 1.log
```
Comment (veezhang) [inline server/README.md]: Here is what to download logs? Why not use the task id nor path?

File: server/README.md [hunk 4]
```diff
@@ -212,31 +334,37 @@ Response:
...
-```
-#### CreateConfigFile API
-Create a config file in config_dir
-```json
-{"config":{"version":"v2","description":"web console import","clientSettings":{"retry":3,"concurrency":10,"channelBufferSize":128,"space":"basketballplayer","connection":{"user":"root","password":"123","address":"192.168.8.145:9669"}},"logPath":"E:\\NebulaProject\\test\\nebula-studio\\server\\data\\tasks/import.log","files":[]},"mountPath":"E:\\NebulaProject\\test\\nebula-studio\\server\\data\\tasks"}
+    "code": 0,
+    "message": "",
+    "data": [
+        "bbbbbbbbbbb",
+        "ccccccccccc"
```
Comment (veezhang) [inline server/README.md]: What is this?

File: server/README.md [hunk 5]
```diff
@@ -266,9 +394,51 @@ Response:
     "code": 0,
     "message": "",
     "data": {
-        "taskDir": "E:\\NebulaProject\\test\\nebula-studio\\server\\data\\tasks",
         "uploadDir": "E:\\NebulaProject\\test\\nebula-studio\\server\\data\\upload"
```
Comment (veezhang) [inline server/README.md:337]: Why expose this?

File: server/README.md [hunk 6]
```diff
@@ -266,9 +394,51 @@ Response:
...
+#### GetTaskDir API
+
+Response:
+
+```json
+{
+    "code": 0,
+    "message": "",
+    "data": {
+        "taskDir": "E:\\NebulaProject\\nebula-studio\\server\\data\\tasks\\1"
```
Comment (veezhang) [inline server/README.md:351]: Why expose this?

File: server/README.md [hunk 7]
```diff
@@ -266,9 +394,51 @@ Response:
...
+Response:
+
+```json
+{
+    "code": 0,
+    "message": "",
+    "data": [
+        {
+            "name": "import.log",
+            "path": "E:\\NebulaProject\\nebula-studio\\server\\data\\tasks\\1\\import.log"
```
Comment (veezhang) [inline server/README.md]: Why expose this?

File: server/config/example-config.yaml [hunk 8]
```diff
@@ -4,4 +4,4 @@ web:
 #  tasks_dir:
 #  sqlitedb_file_path:
 #  ip:
-  port: 7001
\ No newline at end of file
+#  port: 7001
```
Comment (veezhang) [inline server/config/example-config.yaml]: Why comment it ?

File: server/go.mod [hunk 9]
```diff
@@ -3,14 +3,20 @@ module github.com/vesoft-inc/nebula-studio/server
...
 	github.com/vesoft-inc/nebula-http-gateway/ccore v0.0.0-20220124093655-f7f5b4c63e6a
-	github.com/vesoft-inc/nebula-importer v1.0.1-0.20211213064541-05a8646be295
+	github.com/vesoft-inc/nebula-importer v1.0.1-0.20220126025644-1c83ca83aa4f
 	go.uber.org/zap v1.17.0
 	gopkg.in/yaml.v2 v2.4.0
+	gorm.io/driver/sqlite v1.2.6
+	gorm.io/gorm v1.22.5
 )
 
+replace github.com/vesoft-inc/nebula-importer v1.0.1-0.20220126025644-1c83ca83aa4f => github.com/xjlgod/nebula-importer v1.0.1-0.20220311091903-51d03596d67a
```
Comment (veezhang) [inline server/go.mod]: Don't forget to modify here.

File: server/pkg/webserver/controller/files.go [hunk 10]
```diff
@@ -1,18 +1,23 @@
...
 import (
+	"bytes"
 	"encoding/csv"
 	"fmt"
+	"go.uber.org/zap"
 	"io/ioutil"
+	"mime/multipart"
 	"net/http"
 	"os"
 	"path/filepath"
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: sort it by group?

File: server/pkg/webserver/controller/files.go [hunk 11]
```diff
@@ -95,30 +96,90 @@ func FilesIndex(ctx iris.Context) base.Result {
...
+	file, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	reader := bytes.NewReader(file)
+	if err != nil {
+		return err
+	}
+	decoder := mahonia.NewDecoder(charSet)
+	otherBytes, err := ioutil.ReadAll(reader)
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: Is there a problem here if it's a large file?

File: server/pkg/webserver/controller/files.go [hunk 12]
```diff
@@ -95,30 +96,90 @@ func FilesIndex(ctx iris.Context) base.Result {
...
+	}
+	detector := chardet.NewTextDetector()
+	best, err := detector.DetectBest(bytes)
+	if err != nil {
+		return "", err
+	}
+	return best.Charset, nil
+}
 
+func changeFileCharset2UTF8(path string, charSet string) error {
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: `path` is a system package name, maybe other names are better.

File: server/pkg/webserver/controller/gateway.go [hunk 13]
```diff
@@ -1,29 +1,37 @@
...
+	"github.com/vesoft-inc/nebula-studio/server/pkg/utils"
 	"github.com/vesoft-inc/nebula-studio/server/pkg/webserver/base"
 	"github.com/vesoft-inc/nebula-studio/server/pkg/webserver/service/importer"
 
 	"github.com/kataras/iris/v12"
 	"go.uber.org/zap"
 )
 
+const (
+	ErrDir = "err"
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go]: This looks weird.

File: server/pkg/webserver/controller/gateway.go [hunk 14]
```diff
@@ -66,16 +79,35 @@ func ExecNGQL(ctx iris.Context) base.Result {
...
+
+	decode, err := base64.StdEncoding.DecodeString(tokenSlice[1])
+	if err != nil {
+		return base.Response{
+			Code:    base.AuthorizationError,
+			Message: err.Error(),
+		}
+	}
+	account := strings.Split(string(decode), ":")
+	username, password := account[0], account[1]
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go:86]: How about the length of account is less than 2?

File: server/pkg/webserver/controller/gateway.go [hunk 15]
```diff
@@ -115,32 +147,95 @@ func DisconnectDB(ctx iris.Context) base.Result {
...
-	} else {
-		err = importer.Import(taskId, params.ConfigPath, params.ConfigBody)
+		return base.Response{
+			Code:    base.Error,
+			Message: err.Error(),
+		}
 	}
-
+	err = validImportDataParams(params)
+	if err != nil {
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go]: Change it to this?
```go
if err = validImportDataParams(params); err != nil {
    ...
}
```

File: server/pkg/webserver/controller/gateway.go [hunk 16]
```diff
@@ -115,32 +147,95 @@ func DisconnectDB(ctx iris.Context) base.Result {
...
+		zap.L().Warn("importDataParams get fail", zap.Error(err))
+		return base.Response{
+			Code:    base.Error,
+			Message: err.Error(),
+		}
+	}
+	// create config file
+	taskDir := importer.GetNewTaskDir()
+	err = importer.CreateConfigFile(taskDir, *params.ConfigBody)
+	if err != nil {
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go]: go style? Seems like there are more like that.

File: server/pkg/webserver/controller/import.go [hunk 17]
```diff
@@ -62,90 +84,136 @@ func ReadLog(ctx iris.Context) base.Result {
...
 	}
-	if len(bytes) == 0 && taskIdJSON[taskId]{
+	if len(lines) == 0 && taskIdJSON[taskId] {
 		return base.Response{
 			Code: base.Success,
 			Data: "",
 		}
 	}
-	if len(bytes) == 0 {
+	if len(lines) == 0 {
```
Comment (veezhang) [inline server/pkg/webserver/controller/import.go:309]: No logs here, or is there an error?

File: server/pkg/webserver/controller/import.go [hunk 18]
```diff
@@ -62,90 +84,136 @@ func ReadLog(ctx iris.Context) base.Result {
...
 	return base.Response{
 		Code: base.Success,
+		Data: lines,
+	}
+}
+
+func readFile(path string, offset int64, limit int64) ([]string, error) {
+	file, err := os.Open(path)
+	defer file.Close()
+	if err != nil {
```
Comment (veezhang) [inline server/pkg/webserver/controller/import.go]: It's need to close the file when open failed?

File: server/pkg/webserver/service/importer/importer.go [hunk 19]
```diff
@@ -3,10 +3,16 @@ package importer
 import (
 	"errors"
 	"fmt"
+	"gopkg.in/yaml.v2"
+	"os"
+	"path/filepath"
+	"sort"
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: sort it by group?

File: server/pkg/webserver/service/importer/importer.go [hunk 20]
```diff
@@ -22,15 +28,46 @@ type ImportResult struct {
...
+	}
+	path := filepath.Join(dir, fileName)
+	// erase user information
+	address := *config.NebulaClientSettings.Connection.Address
+	user := *config.NebulaClientSettings.Connection.User
+	password := *config.NebulaClientSettings.Connection.Password
+	*config.NebulaClientSettings.Connection.Address = ""
+	*config.NebulaClientSettings.Connection.User = ""
+	*config.NebulaClientSettings.Connection.Password = ""
+	outYaml, err := yaml.Marshal(config)
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go:73]: Why not check the error ?

File: server/pkg/webserver/service/importer/importer.go [hunk 21]
```diff
@@ -22,15 +28,46 @@ type ImportResult struct {
...
+	err = os.WriteFile(path, outYaml, 0644)
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: Why not check the error immediately?

File: server/pkg/webserver/service/importer/importer.go [hunk 22]
```diff
@@ -42,43 +79,59 @@ func Import(taskID string, configPath string, configBody *importconfig.YAMLConfi
...
 		result.TimeCost = fmt.Sprintf("%dms", timeCost)
-
-		if rerr := task.GetRunner().Error(); rerr != nil {
-			// task err: import task not finished err handle
-			task.TaskStatus = StatusAborted.String()
-
-			err, _ := rerr.(importerErrors.ImporterError)
+		if rerrs := task.GetRunner().Errors(); len(rerrs) != 0 {
+			if len(rerrs) == 1 {
+				err, _ := rerrs[0].(importerErrors.ImporterError)
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: Does it must be `importerErrors.ImporterError`?

Also has `err, _ := rerrs[0].(importerErrors.ImporterError)` below.

File: server/pkg/webserver/service/importer/importer.go [hunk 23]
```diff
@@ -42,43 +79,59 @@ func Import(taskID string, configPath string, configBody *importconfig.YAMLConfi
...
 		result.TaskId = taskID
 		result.TimeCost = fmt.Sprintf("%dms", timeCost)
-
-		if rerr := task.GetRunner().Error(); rerr != nil {
-			// task err: import task not finished err handle
-			task.TaskStatus = StatusAborted.String()
-
-			err, _ := rerr.(importerErrors.ImporterError)
+		if rerrs := task.GetRunner().Errors(); len(rerrs) != 0 {
+			if len(rerrs) == 1 {
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: What is the difference between ` == 1`  and `> 1` ?
Just judging `NotCompleteError`? 
Does it not require judgment when `> 1` ?

File: server/pkg/webserver/service/importer/importer.go [hunk 24]
```diff
@@ -42,43 +79,59 @@ func Import(taskID string, configPath string, configBody *importconfig.YAMLConfi
...
 	return nil
 }
 
+func ImportStatus(taskID string) (result *TaskInfo, err error) {
+	if t, ok := GetTaskMgr().GetTask(taskID); ok {
+		if t.GetRunner() != nil {
+			GetTaskMgr().UpdateTaskInfo(taskID)
+		}
+		result = t.TaskInfo
+		return result, nil
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: Why not just return `t.TaskInfo, nil`?

File: server/pkg/webserver/service/importer/importer.go [hunk 25]
```diff
@@ -42,43 +79,59 @@ func Import(taskID string, configPath string, configBody *importconfig.YAMLConfi
...
+	} else {
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: useless `else` ?

File: server/pkg/webserver/service/importer/importer.go [hunk 26]
```diff
@@ -102,19 +157,24 @@ func ImportAction(taskID string, taskAction TaskAction) (result ActionResult, er
 	return result, err
 }
 
+func actionDel(taskID string, result *ActionResult) {
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go:201]: Why return data through input parameters?

File: server/pkg/webserver/service/importer/taskmgr.go [hunk 27]
```diff
@@ -1,130 +1,160 @@
...
 */
-func (mgr *TaskMgr) DelTask(taskID string) {
+func (mgr *TaskMgr) FinishTask(taskID string) {
 	task, ok := mgr.getTaskFromMap(taskID)
-
 	if !ok {
 		return
 	}
-
+	task.UpdateQueryStats()
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/taskmgr.go]: If the update fails, is the finish also done?

Here have similar in `func (mgr *TaskMgr) UpdateTaskInfo(taskID string)`.

File: server/pkg/webserver/service/importer/taskdb.go [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+func (t *TaskDb) InsertTaskInfo(info *TaskInfo) {
+	t.Create(info)
+}
+
+func (t *TaskDb) UpdateTaskInfo(info *TaskInfo) {
+	t.Model(&TaskInfo{}).Where("id = ?", info.ID).Updates(info)
+}
+
+func (t *TaskDb) DelTaskInfo(ID int) {
+	t.Delete(&TaskInfo{}, ID)
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/taskdb.go]: Is it always success here? The other functions are similar. And the caller need to judge the error?

File: server/pkg/webserver/service/importer/taskdb.go [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+func (t *TaskDb) LastId() int {
+	var id int
+	if t.Raw("SELECT MAX(id) FROM task_infos").Scan(&id).Error != nil {
+		return 0
+	}
+	return id
+}
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/taskdb.go]: Is it need to return an error? Using `1`(`0 + 1`) as the `id` when `error != nil`?

File: server/pkg/webserver/service/importer/taskmgr.go [hunk 30]
```diff
@@ -1,130 +1,160 @@
...
-func NewTaskID() (taskID string) {
+func (mgr *TaskMgr) NewTask(nebulaAddress string, user string, name string, space string) (*Task, string, error) {
 	mux.Lock()
 	defer mux.Unlock()
-	tid++
-	taskID = fmt.Sprintf("%d", tid)
-	return taskID
+	task := newTask(nebulaAddress, user, name, space)
+	mgr.db.InsertTaskInfo(task.TaskInfo)
+	tid := mgr.db.LastId()
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/taskmgr.go]: The `id` is not necessarily the `id` of the task created above.

File: server/pkg/webserver/service/importer/taskmgr.go [hunk 31]
```diff
@@ -133,69 +163,25 @@ func (mgr *TaskMgr) StopTask(taskID string) bool {
...
-
-	sqlStmt := `
-		create table tasks (taskID integer not null primary key, taskStatus text, taskMessage text);
-		delete from tasks;
-	`
-	_, err = GetTaskMgr().db.Exec(sqlStmt)
-
-	if err != nil {
-		zap.L().Fatal(fmt.Sprintf("%q: %s\n", err, sqlStmt))
+	for i := 1; i <= mgr.db.LastId(); i++ {
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/taskmgr.go]: Why not select it from db?

File: server/pkg/api/routes/import.go [hunk 32]
```diff
@@ -11,11 +11,11 @@ var ImportRoute = base.Route{
 	SubRoutes: []base.Route{
 		{
 			Path: "log",
-			GET:  controller.ReadLog,
+			GET:  controller.ReadImportLog,
 		},
 		{
-			Path: "config",
-			POST: controller.CreateConfigFile,
+			Path: "err_log",
```
Comment (veezhang) [inline server/pkg/api/routes/import.go]: Use `-` nor `_` in http path?

File: server/pkg/utils/file.go [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+
+func CreateDir(dir string) error {
+	_, err := os.Stat(dir)
+	if os.IsNotExist(err) {
+		err := os.MkdirAll(dir, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```
Comment (veezhang) [inline server/pkg/utils/file.go:12]: The `err` may not be `nil`.

File: server/pkg/webserver/controller/files.go [hunk 34]
```diff
@@ -95,30 +95,88 @@ func FilesIndex(ctx iris.Context) base.Result {
...
 }
 
 func FilesUpload(ctx iris.Context) base.Result {
-
 	dir := config.Cfg.Web.UploadDir
 	files, _, err := ctx.UploadFormFiles(dir)
+	for _, file := range files {
+		charSet, err := checkCharset(file)
+		if err != nil {
+			continue
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: It's need to return `err`?

File: server/pkg/webserver/controller/files.go [hunk 35]
```diff
@@ -95,30 +95,88 @@ func FilesIndex(ctx iris.Context) base.Result {
...
+	decoder := mahonia.NewDecoder(charSet)
+	otherBytes, err := ioutil.ReadAll(reader)
+	s := string(otherBytes)
+	utf8String := decoder.ConvertString(s)
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: dup?

File: server/pkg/webserver/controller/files.go [hunk 36]
```diff
@@ -95,30 +95,88 @@ func FilesIndex(ctx iris.Context) base.Result {
...
+	err = ioutil.WriteFile(path, []byte(utf8String), 0666)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (veezhang) [inline server/pkg/webserver/controller/files.go]: You can just return `ioutil.WriteFile(path, []byte(utf8String), 0666)`.

File: server/pkg/webserver/controller/gateway.go [hunk 37]
```diff
@@ -168,3 +263,45 @@ func HandleImportAction(ctx iris.Context) base.Result {
...
+		return base.Response{
+			Code:    base.Error,
+			Message: "id parse failed",
+		}
+	}
+	configPath := filepath.Join(config.Cfg.Web.TasksDir, id, "config.yaml")
+	ctx.SendFile(configPath, "config.yaml")
+	return base.Response{
+		Code: base.Success,
+	}
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go]: Is there any abnormality call `ctx.JSON(&result)` after `ctx.SendFile` ?

File: server/pkg/webserver/controller/gateway.go [hunk 38]
```diff
@@ -168,3 +263,45 @@ func HandleImportAction(ctx iris.Context) base.Result {
...
+		return base.Response{
+			Code:    base.Error,
+			Message: "path parse failed",
+		}
+	}
+	index := strings.LastIndex(path, "/")
+	if index == -1 {
+		index = strings.LastIndex(path, "\\")
+	}
+	filename := path[index:]
```
Comment (veezhang) [inline server/pkg/webserver/controller/gateway.go]: How about `index == -1`?

File: server/pkg/webserver/controller/import.go [hunk 39]
```diff
@@ -62,90 +84,136 @@ func ReadLog(ctx iris.Context) base.Result {
...
+		zap.L().Warn("open file error", zap.Error(err))
+		return nil, err
+	}
+	scanner := bufio.NewScanner(file)
```
Comment (veezhang) [inline server/pkg/webserver/controller/import.go:410]: Wouldn't that be more concise?
```go
	for lineIndex := int64(0); scanner.Scan() && lineIndex < offset+limit; lineIndex++ {
		if lineIndex >= offset {
			res = append(res, scanner.Text())
		}
	}
```

File: server/README.md (hunk 3 above)
Comment (xjlgod) [reply server/README.md]: Because there is need to download two types of files with same id.

File: server/README.md (hunk 5 above)
Comment (xjlgod) [reply server/README.md:337]: To get all files from this dir

File: server/config/example-config.yaml (hunk 8 above)
Comment (xjlgod) [reply server/config/example-config.yaml]: It is the port in production enviroment.

File: server/pkg/api/routes/import.go (hunk 32 above)
Comment (xjlgod) [reply server/pkg/api/routes/import.go]: ok

File: server/README.md (hunk 1 above)
Comment (xjlgod) [reply server/README.md]: Ok, I will modify it

File: server/README.md (hunk 6 above)
Comment (xjlgod) [reply server/README.md:351]: To get different files from this dir

File: server/README.md (hunk 7 above)
Comment (xjlgod) [reply server/README.md]: To download this file

File: server/pkg/utils/file.go (hunk 33 above)
Comment (xjlgod) [reply server/pkg/utils/file.go:12]: Ok, I hava fixed it

File: server/pkg/webserver/controller/files.go (hunk 36 above)
Comment (xjlgod) [reply server/pkg/webserver/controller/files.go]: Ok

File: server/pkg/webserver/controller/gateway.go (hunk 38 above)
Comment (xjlgod) [reply server/pkg/webserver/controller/gateway.go]: The path may be under linux or windows env.

File: server/pkg/webserver/controller/import.go (hunk 17 above)
Comment (xjlgod) [reply server/pkg/webserver/controller/import.go:309]: If no logs here and the task is not existed, it is an error.

File: server/pkg/webserver/controller/import.go (hunk 18 above)
Comment (xjlgod) [reply server/pkg/webserver/controller/import.go]: I have fixed it

File: server/pkg/webserver/service/importer/importer.go (hunk 23 above)
Comment (xjlgod) [reply server/pkg/webserver/service/importer/importer.go]: We want to ignore the NotCompleteError.

File: server/pkg/webserver/service/importer/taskdb.go (hunk 28 above)
Comment (xjlgod) [reply server/pkg/webserver/service/importer/taskdb.go]: I have fixed it

File: server/pkg/webserver/service/importer/taskdb.go (hunk 29 above)
Comment (xjlgod) [reply server/pkg/webserver/service/importer/taskdb.go]: If there is no data in the table, an error will occur here. The self incrementing ID will not be 0.

File: server/pkg/webserver/service/importer/taskmgr.go (hunk 31 above)
Comment (xjlgod) [reply server/pkg/webserver/service/importer/taskmgr.go]: A task may have been deleted.

File: server/README.md (hunk 3 above)
Comment (veezhang) [reply server/README.md]: The client should not know the path of the service.
You can use parameters and paths to differentiate.

File: server/README.md (hunk 5 above)
Comment (veezhang) [reply server/README.md:337]: https://github.com/vesoft-inc/nebula-studio/pull/117#discussion_r826643859

File: server/README.md (hunk 6 above)
Comment (veezhang) [reply server/README.md:351]: https://github.com/vesoft-inc/nebula-studio/pull/117#discussion_r826643859

File: server/README.md (hunk 7 above)
Comment (veezhang) [reply server/README.md]: https://github.com/vesoft-inc/nebula-studio/pull/117#discussion_r826643859

File: server/config/example-config.yaml (hunk 8 above)
Comment (veezhang) [reply server/config/example-config.yaml]: I thank should not be commented out if it's the default port for production environment.
Why comment it? for testing , local development or ?

File: server/pkg/webserver/controller/gateway.go (hunk 14 above)
Comment (veezhang) [reply server/pkg/webserver/controller/gateway.go:86]: Forget?

File: server/pkg/webserver/controller/import.go (hunk 17 above)
Comment (veezhang) [reply server/pkg/webserver/controller/import.go:309]: Why not check for existence first?

File: server/pkg/webserver/service/importer/importer.go (hunk 20 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/importer.go:73]: Forget?

File: server/pkg/webserver/service/importer/importer.go [hunk 40]
```diff
@@ -22,15 +28,48 @@ type ImportResult struct {
...
+	err = os.WriteFile(path, outYaml, 0644)
+	if err != nil {
+		zap.L().Warn("write"+path+"file error", zap.Error(err))
+		return err
+	}
+	defer func() {
+		*config.NebulaClientSettings.Connection.Address = address
+		*config.NebulaClientSettings.Connection.User = user
+		*config.NebulaClientSettings.Connection.Password = password
+	}()
```
Comment (veezhang) [inline server/pkg/webserver/service/importer/importer.go]: It's need `defer`? It's already at the end of func.

File: server/pkg/webserver/service/importer/importer.go (hunk 23 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/importer.go]: If the length of `rerrs` gt `0`, it's need not to check it?

File: server/pkg/webserver/service/importer/importer.go (hunk 26 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/importer.go:201]: Forget?

File: server/pkg/webserver/service/importer/taskmgr.go (hunk 27 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/taskmgr.go]: Forget?

File: server/pkg/webserver/service/importer/taskdb.go (hunk 29 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/taskdb.go]: So why you return `0`?

File: server/pkg/webserver/service/importer/taskmgr.go (hunk 30 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/taskmgr.go]: Forget?

File: server/pkg/webserver/service/importer/taskmgr.go (hunk 31 above)
Comment (veezhang) [reply server/pkg/webserver/service/importer/taskmgr.go]: Why the user need to get the deleted task?

File: server/pkg/webserver/service/importer/importer.go (hunk 40 above)
Comment (xjlgod) [reply server/pkg/webserver/service/importer/importer.go]: If err happen，it also need to do

File: server/pkg/webserver/controller/gateway.go (hunk 37 above)
Comment (veezhang) [reply server/pkg/webserver/controller/gateway.go]: Forget?

File: server/pkg/webserver/controller/gateway.go (hunk 38 above)
Comment (veezhang) [reply server/pkg/webserver/controller/gateway.go]: I means that the value of `index` maybe `-1` here, so it will panic.

**Review by veezhang — APPROVED** [review]
LGTM

**Review by veezhang — APPROVED** [review]
LGTM

**Pull request merged**